### DISCUSSION TITLE
Cookies sent back from context but not mutation

### DISCUSSION THREAD

**Discussion by ahriannn** [trpc/trpc#4800] [Q&A] [answered]
Hi there,

Hoping someone can understand this. I'm using an express server w/ express-session & TRPC. When i set `req.session` inside my createContext, the cookie is sent back with session, however if i do it in my mutation, nothing is sent back to the user. This is also the case if I did `res.cookie`

Anyone understand why?

Cookie sent back:
```
export const createTRPCContext = ({
  req,
  res,
}: CreateExpressContextOptions) => {
  req.session.user = {
    id: "1",
  };

  return {
    db,
    session: {
      user: req.session.user,
    },
    req,
    res,
  };
};
```

Cookie not sent:
```
 authUser: publicProcedure
    .input(
      z.object({
        email: z.string().email(),
        password: z.string(),
      }),
    )
    .mutation(async ({ ctx, input }) => {
      ctx.req.session.user = {
        id: "1"
      };

      return {};
    }),
```

**Comment by ahriannn** [comment]
Added note; when setting the session inside the procedure, I'm seeing it being made inside my session store, so it's purely the mutation not sending back the headers...

**Comment by ahriannn** [comment, marked as answer]
Issue fixed, it's due to using `unstable_httpBatchStreamLink` which streams headers at the start of the request & changing it to `httpBatchLink` in my client, which will mean it sends them in the end as far as i'm aware.

@KATT I think it should be documented that if you require sending cookies etc via mutations to not use `unstable_httpBatchStreamLink` as things such as T3 stack use `unstable_httpBatchStreamLink` by default and I'm sure has caused some issues for others...

**Reply by KATT** [reply]
Please click the edit button and add an appropriate note about it 🙏

**Reply by lovrozagar** [reply]
Bless you sir, setting a cookie with httpBatchLink works.
Noting that T3 app still uses unstable_httpBatchStreamLink.

**Reply by madebydor** [reply]
> Bless you sir, setting a cookie with httpBatchLink works. Noting that T3 app still uses unstable_httpBatchStreamLink.

Gosh darn it, I wasted HOURS not understanding why I can't create cookies
I tried all ways!
````
ctx.headers.set("Set-Cookie", sessionCookie.serialize()); // didn't work
cookies().set("name", "lee"); // didn't work
````

But nothing worked until I came across this!

What is the impact of changing `unstable_httpBatchStreamLink` to `httpBatchLink`?
